src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriter.cs

```diff
@@ -51,5 +51,44 @@ public virtual ValueTask<FlushResult> WriteAsync(ReadOnlyMemory<byte> source, Ca
             this.Write(source.Span);
             return FlushAsync(cancellationToken);
         }
+
+        /// <summary>
+        /// Asynchronously reads the bytes from the specified stream and writes them to the <see cref="PipeWriter"/>.
+        /// </summary>
+        /// <param name="source">The stream from which the contents will be copied.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>A task that represents the asynchronous copy operation.</returns>
+        protected internal virtual async Task CopyFromAsync(Stream source, CancellationToken cancellationToken = default)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                Memory<byte> buffer = GetMemory();
+                int read = await source.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
+
+                if (read == 0)
+                {
+                    break;
+                }
+
+                Advance(read);
+
+                FlushResult result = await FlushAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Is it important to flush on every Advance?  Or maybe the pipe doesn't actually need to flush in response to this being requested and can leverage its own heuristics as to when to actually do so?

Reply:
You could imagine batching but that would complicate things and it's something we'd do as an optimization after measuring and deciding it makes sense to do so. This is the simplest implementation.

Reply:
Ok